### PR TITLE
[#144] Add `traverseToSnd` and friends to `Relude.Extra.Tuple`

### LINKED ISSUE #144: Add `traverseToSnd` and friends to `Relude.Extra.Tuple`
I think `traverse` analogues to the `map` functions in `Relude.Extra.Tuple` would be useful:

```haskell
traverseToFst :: Functor t => (a -> t b) -> a -> t (b, a)
traverseToFst f a = flip (,) a <$> f a

traverseToSnd ::  Functor t => (a -> t b) -> a -> t (a, b)
traverseToSnd f a = (,) a <$> f a

traverseBoth :: Applicative t => (a -> t b) -> (a, a) -> t (b, b)
traverseBoth f (a1, a2) = (,) <$> f a1 <*> f a2
```

(No analogue of `dupe`, at least for now, because there'd be ambiguity over whether it worked like `\a -> join (,) <$> a` or like `\a -> (,) <$> a <*> a`, and I'm not sure which would be better.)

The use case that inspired this was a function (`FilePath -> IO (FilePath, String)`) that takes a filename, and returns a pair of the name and the contents. With this proposal, such a function would simply be `traverseToSnd readFile`.

If this idea is acceptable, I'll make a PR that adds it along with hlint rules and tests.

### PR DESCRIPTION
**Pull request by josephcsible**
Resolves #144

### HLint

- [x] I've changed the exposed interface (add new reexports, remove reexports, rename reexported things, etc.).
  - [x] I've updated [`hlint.dhall`](https://github.com/kowainik/relude/blob/master/hlint/hlint.dhall) accordingly to my changes (add new rules for the new imports, remove old ones, when they are outdated, etc.).
  - [x] I've generated the new `.hlint.yaml` file (see [this instructions](https://github.com/kowainik/relude#generating-hlintyaml)).

### General

- [x] I've updated the [CHANGELOG](https://github.com/kowainik/relude/blob/master/CHANGELOG.md) with the short description of my latest changes.
- [x] All new and existing tests pass.
- [x] I keep the code style used in the files I've changed (see [style-guide](https://github.com/kowainik/org/blob/master/style-guide.md#haskell-style-guide) for more details).
- [x] I've used the [`stylish-haskell` file](https://github.com/kowainik/relude/blob/master/.stylish-haskell.yaml).
- [x] My change requires the documentation updates.
  - [x] I've updated the documentation accordingly.
- [ ] I've added the `[ci skip]` text to the docs-only related commit's name.

### CODE REVIEW

File: hlint/hlint.dhall [hunk 1]
```diff
@@ -882,4 +882,18 @@ in [ rule.Arguments { arguments =
...
+
+   -- Tuple
+   , hintNote "fmap (,a) (f a)"         "traverseToFst f a" "Use `traverseToFst`"
+   , hintNote "fmap (flip (,) a) (f a)" "traverseToFst f a" "Use `traverseToFst`"
+   , hintNote "(,a) <$> f a"            "traverseToFst f a" "Use `traverseToFst`"
+   , hintNote "flip (,) a <$> f a"      "traverseToFst f a" "Use `traverseToFst`"
+   , hintNote "fmap (a,) (f a)"    "traverseToSnd f a" "Use `traverseToSnd`"
+   , hintNote "fmap ((,) a) (f a)" "traverseToSnd f a" "Use `traverseToSnd`"
+   , hintNote "(a,) <$> f a"       "traverseToSnd f a" "Use `traverseToSnd`"
+   , hintNote "(,) a <$> f a"      "traverseToSnd f a" "Use `traverseToSnd`"
```
Comment (josephcsible) [inline hlint/hlint.dhall]: I can't think of any non-contrived rules for `traverseBoth`, especially ones that wouldn't involve `curry` or something equivalent to it.

File: src/Relude/Extra/Tuple.hs [hunk 2]
```diff
@@ -11,10 +13,12 @@ module Relude.Extra.Tuple
...
        , mapToSnd
        , mapBoth
+       , traverseToFst
+       , traverseToSnd
+       , traverseBoth
        ) where
 
--- $setup
--- >>> import Relude
+import Relude
```
Comment (josephcsible) [inline src/Relude/Extra/Tuple.hs]: I did this to bring `Functor` and `Applicative` into scope.

**Comment by josephcsible** [comment]
I ended up not adding an analogue to `dupe`, since the version that doesn't perform the effects twice would just be equivalent to `fmap dupe`, which I think is too simple to be worth adding.

File: hlint/hlint.dhall [hunk 3]
```diff
@@ -882,4 +882,18 @@ in [ rule.Arguments { arguments =
    , warnLifted "putBSLn" "x"
    , warnLifted "putLBS" "x"
    , warnLifted "putLBSLn" "x"
+
+   ------------
+   -- Extra
+   ------------
+
+   -- Tuple
+   , hintNote "fmap (,a) (f a)"         "traverseToFst f a" "Use `traverseToFst`"
```
Comment (josephcsible) [inline hlint/hlint.dhall]: How do I make this aware of the need to `import Relude.Extra.Tuple`?

File: src/Relude/Extra/Tuple.hs [hunk 4]
```diff
@@ -60,3 +64,47 @@ mapToSnd f a = (a, f a)
...
+traverseToFst :: Functor t => (a -> t b) -> a -> t (b, a)
+traverseToFst f a = (,a) <$> f a
+{-# INLINE traverseToFst #-}
+
+{- | Apply a function that returns a value inside of a functor,
+with the output in the second slot, the input in the first,
+and the entire tuple inside the functor.
+
+A dual to 'traverseToFst'.
+
```
Comment (josephcsible) [inline src/Relude/Extra/Tuple.hs]: > functions that have type like `FilePath -> IO (FilePath, String)` seems valuable, I think this text also can be added to the documentation 📚

I can't think of good wording for this. Any advice?

File: hlint/hlint.dhall (hunk 3 above)
Comment (ChShersh) [reply hlint/hlint.dhall]: What do you mean here? You want to tell users that `traverseToFst` comes from the `Relude.Extra.Tuple` module? In that case you can just adjust the message of the note:

```
Use `traverseToFst` from `Relude.Extra.Tuple`
```

File: hlint/hlint.dhall (hunk 1 above)
Comment (ChShersh) [reply hlint/hlint.dhall]: Existing examples are already good enough :+1: Don't need to support all crazy combinations.

File: src/Relude/Extra/Tuple.hs (hunk 4 above)
Comment (ChShersh) [reply src/Relude/Extra/Tuple.hs]: Hmm.. I see that the existing documentation is already good enough. And examples are nice. Probably don't need to tell more here.

**Comment by josephcsible** [comment]
> But it would be nice if you can note about `Relude.Extra.Tuple` to the HLint rules.

Note about that where? In the changelog?

File: hlint/hlint.dhall (hunk 3 above)
Comment (josephcsible) [reply hlint/hlint.dhall]: I mean a way to make hlint itself aware of the required import, so if someone were to run it with `--refactor`, it wouldn't break their code by adding a usage of it without the import. Is this possible? If not, then I will just add it to the message for the user.

**Comment by ChShersh** [comment]
@josephcsible I mean in every HLint rule. Instead of:

```
   , hintNote "fmap (,a) (f a)"         "traverseToFst f a" "Use `traverseToFst`"
```

you can write

```
   , hintNote "fmap (,a) (f a)"         "traverseToFst f a" "Use `traverseToFst` from `Relude.Extra.Tuple`"
```

File: hlint/hlint.dhall (hunk 3 above)
Comment (ChShersh) [reply hlint/hlint.dhall]: > Is this possible? If not, then I will just add it to the message for the user.

I think it's not possible. It's one of the reasons why it's just a hint, not a warning. So adding just message is okay 👍 

Note, that if you're using `base-noprelude` to add custom relude, you only need to add this import in one place for your application.

**Comment by josephcsible** [comment]
Okay, I think this is ready to merge now.

**Pull request merged**